File: bridge/routes/stepmotor.py
from flask import Blueprint, jsonify, request

from bridge.modules.stepmotor_mod import stepmotor
from bridge.shared.locks import stepmotor_lock


stepmotor_bp = Blueprint("stepmotor", __name__, url_prefix="/stepmotor")


@stepmotor_bp.route("/turn_degrees", methods=["POST", "OPTIONS"])
def turn_degrees():
    if request.method == "OPTIONS":
        return "", 204

    data = request.get_json(silent=True) or {}
    degrees = data.get("degrees", 0)
    direction = str(data.get("direction", "cw"))
    speed = data.get("speed", 1.0)

    try:
        degrees = float(degrees)
    except (TypeError, ValueError):
        degrees = 0.0

    if direction not in ("cw", "ccw"):
        direction = "cw"

    try:
        speed = float(speed)
    except (TypeError, ValueError):
        speed = 1.0

    with stepmotor_lock:
        stepmotor.turn_degrees(degrees, direction=direction, speed=speed)

    return jsonify(ok=True)


@stepmotor_bp.route("/turn_steps", methods=["POST", "OPTIONS"])
def turn_steps():
    if request.method == "OPTIONS":
        return "", 204

    data = request.get_json(silent=True) or {}
    steps = data.get("steps", 0)
    direction = str(data.get("direction", "cw"))
    speed = data.get("speed", 1.0)

    try:
        steps = int(steps)
    except (TypeError, ValueError):
        steps = 0

    if direction not in ("cw", "ccw"):
        direction = "cw"

    try:
        speed = float(speed)
    except (TypeError, ValueError):
        speed = 1.0

    with stepmotor_lock:
        stepmotor.turn_steps(steps, direction=direction, speed=speed)

    return jsonify(ok=True)


@stepmotor_bp.route("/reset_position", methods=["POST", "OPTIONS"])
def reset_position():
    if request.method == "OPTIONS":
        return "", 204

    with stepmotor_lock:
        stepmotor.reset_position()

    return jsonify(ok=True)


@stepmotor_bp.route("/get_position", methods=["GET"])
def get_position():
    with stepmotor_lock:
        degrees = stepmotor.get_position_degrees()

    return jsonify(ok=True, degrees=round(float(degrees), 1))

File: extension/stepmotor.js
(function (Scratch) {
  'use strict';

  class CrowPiStepMotor {
    getInfo () {
      return {
        id: 'crowpistepmotor',
        name: 'CrowPi Step Motor',
        color1: '#8D6E63',
        color2: '#795548',
        color3: '#5D4037',
        blocks: [
          {
            opcode: 'turnDegrees',
            blockType: Scratch.BlockType.COMMAND,
            text: 'tourner le moteur pas à pas de [DEGREES] ° [DIR] à vitesse [SPEED]',
            arguments: {
              DEGREES: {
                type: Scratch.ArgumentType.ANGLE,
                defaultValue: 360
              },
              DIR: {
                type: Scratch.ArgumentType.STRING,
                defaultValue: 'horaire',
                menu: 'direction'
              },
              SPEED: {
                type: Scratch.ArgumentType.NUMBER,
                defaultValue: 5
              }
            }
          },
          {
            opcode: 'turnSteps',
            blockType: Scratch.BlockType.COMMAND,
            text: 'tourner le moteur pas à pas de [STEPS] pas [DIR] à vitesse [SPEED]',
            arguments: {
              STEPS: {
                type: Scratch.ArgumentType.NUMBER,
                defaultValue: 10
              },
              DIR: {
                type: Scratch.ArgumentType.STRING,
                defaultValue: 'horaire',
                menu: 'direction'
              },
              SPEED: {
                type: Scratch.ArgumentType.NUMBER,
                defaultValue: 5
              }
            }
          },
          {
            opcode: 'resetPosition',
            blockType: Scratch.BlockType.COMMAND,
            text: 'réinitialiser la position du moteur pas à pas'
          },
          {
            opcode: 'getPosition',
            blockType: Scratch.BlockType.REPORTER,
            text: 'position logique du moteur pas à pas (°)'
          }
        ],
        menus: {
          direction: {
            acceptReporters: false,
            items: ['horaire', 'anti-horaire']
          }
        }
      };
    }

    _dirToCode (dir) {
      const s = String(dir || '').toLowerCase();
      if (s.includes('anti')) {
        return 'ccw';
      }
      return 'cw';
    }

    async turnDegrees (args) {
      const degrees = Number(args.DEGREES || 0);
      const direction = this._dirToCode(args.DIR);
      const speed = Number(args.SPEED || 1);
      try {
        await fetch('http://127.0.0.1:3232/stepmotor/turn_degrees', {
          method: 'POST',
          headers: {
            'Content-Type': 'application/json'
          },
          body: JSON.stringify({
            degrees,
            direction,
            speed
          })
        });
      } catch (e) {
        // ignore errors
      }
    }

    async turnSteps (args) {
      const steps = Number(args.STEPS || 0);
      const direction = this._dirToCode(args.DIR);
      const speed = Number(args.SPEED || 1);
      try {
        await fetch('http://127.0.0.1:3232/stepmotor/turn_steps', {
          method: 'POST',
          headers: {
            'Content-Type': 'application/json'
          },
          body: JSON.stringify({
            steps,
            direction,
            speed
          })
        });
      } catch (e) {
        // ignore errors
      }
    }

    async resetPosition () {
      try {
        await fetch('http://127.0.0.1:3232/stepmotor/reset_position', {
          method: 'POST',
          headers: {
            'Content-Type': 'application/json'
          }
        });
      } catch (e) {
        // ignore errors
      }
    }

    async getPosition () {
      try {
        const res = await fetch('http://127.0.0.1:3232/stepmotor/get_position');
        const data = await res.json();
        if (data && data.ok && typeof data.degrees === 'number') {
          return data.degrees;
        }
      } catch (e) {
        // ignore errors
      }
      return 0;
    }
  }

  Scratch.extensions.register(new CrowPiStepMotor());
})(Scratch);
